### DISCUSSION TITLE
Issue with Grid view in Airflow 2.3.1rc1

### DISCUSSION THREAD

**Discussion by MatthieuBlais** [apache/airflow#23908] [Q&A] [answered]
We are facing an issue with the Grid view in Airflow 2.3. We would like to know if this issue can be reproduced by others or not. 

- We tried to upgrade our development environment from Airflow 2.4.0 to 2.3.0. We have 200+ Dags and 6M+ records in the task_instance table. We run on K8s, and deployed Airflow with a Helm chart.
- We faced some issues while upgrading the db. It was likely related to this [issue](https://github.com/apache/airflow/issues/23472).
- To work around, we upgraded to version 2.3.1rc1, and we managed to complete the upgrade.
- While testing changes, we noticed some grid views cannot be loaded. When we open the DAG page, the area remains blank

![image](https://user-images.githubusercontent.com/9677913/170163571-55518cdb-0286-4314-874a-096538b38b9c.png)

- Looking at the JS console, we can see some errors. These errors are shown only for DAGs that can't load the grid view.

![image](https://user-images.githubusercontent.com/9677913/170163685-0c44e825-e03c-4c4f-8b5b-8a1f8b99d9f0.png)

- Looking at grid.js, the error happens while filtering an array of instances based on a runId. Some of these instances are null, hence the error.

![image](https://user-images.githubusercontent.com/9677913/170164258-2c19e85c-e4e2-4b8d-84de-3d7cdd61226f.png)

- We initially thought it could be related to our issues while upgrading the db. We could have corrupted some records.
- After more investigation, we noticed the DAG we were looking at had been modified recently (1 task had been added). Looking at our DB, we had 8 records in the task_instance for this new task and 25 records for the other task of this DAG. This matches with what we could see in the array of instances before the filtering. 
- To confirm that, we decided to update another DAG with a grid view shown on the UI. We added a task and deployed it. The error could be reproduced, and the grid view wasn't displayed anymore.

If this bug can be reproduced by others, I believe there are 2 ways to fix this:
- In the JS file, we check the instance is not null before checking the runId value.
- We can fetch the last 25 instances for a given task, but if there are less than 25 records, we should not have null values to fill up the array.

**Comment by zachliu** [comment]
i can reproduce this in `2.3.1`

also i had problems on the db migration when i tried to upgrade from Airflow 2.2.5 to 2.3.0

had to remove quite a lot of "dangling" items after the db migration

i think you should create a github issue on this

**Comment by zachliu** [comment]
ah someone already created one https://github.com/apache/airflow/issues/23588
it seems they have problems reproducing it, this discussion should help them

**Reply by jedcunningham** [reply]
I think that is a different issue. This one, however, is really easy to reproduce:

```python
from datetime import datetime

from airflow import DAG
from airflow.operators.bash import BashOperator

with DAG(
    dag_id="simple", start_date=datetime(2022, 4, 1), schedule_interval=None
) as dag:
    BashOperator(task_id="hello", bash_command="echo hello")
    # BashOperator(task_id="world", bash_command="echo world")
```

Run it once with the second task commented, then uncomment it, wait for the DAG to be parsed again, and it'll blow up.

This was not an issue in 2.3.0, so this is a regression in 2.3.1.

**Reply by jedcunningham** [reply]
I've confirmed this is the PR that caused the regression: https://github.com/apache/airflow/pull/23813

**Reply by jedcunningham** [reply]
I've got a draft PR with a fix open: https://github.com/apache/airflow/pull/23924

**Reply by zachliu** [reply]
simple fix indeed :+1:

**Reply by jedcunningham** [reply]
This is a better fix (#23947) and also is much faster.

**Comment by jedcunningham** [comment, marked as answer]
And #23947, which will be in 2.3.2, fixes this. Thanks everyone.
